THREESCALE-9492: Spam protection refactor

jlledom opened the pull request:
**What this PR does / why we need it**:

The spam protection needed a refactor to meet the requirements in [THREESCALE-9492](https://issues.redhat.com/browse/THREESCALE-9492). The issue requires configuring a different set of checks for the `signup` and the `reset_password` screens in the developer portal. For instance, it's ok to suspect on a client filling the signup form in less than ten seconds, that's too fast, but that same behavior is not suspicious in the reset password form, since it only has one field.

Having different checks for different screens wasn't possible with the old design, because spam checks were attached to models, and both screens mentioned above work over the same model, `Account`, so both had to share the checks.

[THREESCALE-9492](https://issues.redhat.com/browse/THREESCALE-9492) is also a dependency for [THREESCALE-6645](https://issues.redhat.com/browse/THREESCALE-6645), which requires enabling the bot detection for the developer portal login screen, with it's own set of checks.

**Which issue(s) this PR fixes** 

[THREESCALE-9492](https://issues.redhat.com/browse/THREESCALE-9492)

**Verification steps** 

Open the signup or the reset password screen and try to trigger the spam detection:

- Signup screen. Any combination of:
  - Disable javascript.
  - Check the honeypot checkbox with `document.getElementById("#confirmation").checked = true`.
  - Fill in the form very fast. Send the form with wrong data to refresh the form already filled, except password fields, which can be filled quickly with `Ctrl+V`.
  - Send the form three times within a minute.
- Forgot password screen. Any combination of:
  - Disable javascript.
  - Check the honeypot checkbox with `document.getElementById("#confirmation").checked = true`.
  - Send the form three times within a minute.

**Notes for the reviewer**

### Old design
- Business logic:
  - `SessionStore` ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/app/lib/three_scale/spam_protection/session_store.rb)): just a wrapper to read/write from the cookie
    - Methods `mark_possible_spam` and `marked_as_possible_spam?`
    - Read/write a timestamp called `marked_as_possible_spam_until`. Which force showing the captcha if it's set to less than 5 minutes ago. So the captcha expires after five minutes.
  - `Protector` ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/app/lib/three_scale/spam_protection/protector.rb)): is the business logic that checks whether it's spam or not
    - Is attached to a model
    - The main method is `spam?`, which runs the spam challenges. This is called from the controller on `PUT` or `POST` operations, in order to update `marked_as_possible_spam_until`
  - `FormProtector` ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/app/lib/three_scale/spam_protection/protector.rb#L31-L112)): generates the proper html code to show either the captcha or the checks fields
    - The main method is `to_str` which reads `marked_as_possible_spam_until` to decide either to return the check inputs of the captcha.
    - This method is called from `FormBuilder#spam_protection`. Since `SemanticFormBuilder` includes `FormBuilder`, it has `spam_protection` available from templates.
    - The method is also called from `SpamProtected`. From which some Liquid forms inherit, so they also have the spam protection available.
  - `Configuration` ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/app/lib/three_scale/spam_protection/configuration.rb)): the logic that manages the checks for a model
      - `available_checks`: Checks that exist
      - `enabled_checks`: a list of symbols to identify checks that have been initialized and its `apply!` method called, so the model responds to check methods
      - `active_checks`: a list of objects which correspond to the enabled_checks (symbols)
      - It also manages a `store` where checks can store data. Note this store is not the SessionStore attached to the cookie. This another store, merely a class attribute, which is therefore one single hash accessible from all the app, and fortunately unused.
- Integration
  - Models
    - `ThreeScale::SpamProtection::Integration::Model` ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/app/lib/three_scale/spam_protection/integration/model.rb)), to be included from models, which:
      - defines a `spam_protection` instance attribute of type `ThreeScale::SpamProtection::Protector`
      - defines a `spam_protection` class attribute of type `ThreeScale::SpamProtection::Configuration`
      - has a list of spam checks attached to the model class, not instance
      - is configured by calling `has_spam_protection` on the model
        - This calls the `apply!` method on every check, to add attributes to the model class
  - Controllers
    - `ThreeScale::SpamProtection::Integration::Controller` ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/app/lib/three_scale/spam_protection/integration/controller.rb)), to be included from controller, which adds two methods: `spam_check` and `spam_check_save`
      - On `PUT` and `POST` requests, the controller can call one of those to trigger the spam check and set `marked_as_possible_spam_until` is needed.
  - Forms
    - `ThreeScale::SpamProtection::Integration::FormBuilder` ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/app/lib/three_scale/spam_protection/integration/form_builder.rb)):
      - Adds a `spam_protection` method that retrieves the html code from the `FormProtector` object mentioned above
      - This is to build spam protected forms from views like `*.slim` files
    - `Liquid::forms::SpamProtected` ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/lib/developer_portal/lib/liquid/forms/spam_protected.rb)):
      - This is to build spam protected forms from liquid templates.
      - Automatically attaches the html code from `FormProtector`
- Checks
  - A list of spam checks
     - `:javascript`: Passes when javascript is enabled, suspects otherwise.
     - `:honeypot`: Adds an invisible checkbox to the form. It passes when the form is not checked, suspects otherwise.
     - `:timestamp`: Passes when the form takes more than ten seconds to be filled, suspects when it takes less.
   - All checks return a number between `0` and `1` on how likely they think the client is a bot. From all the checks an average is computed, which `if > 0.4`, the client is considered a bot.
   - What needs a check to provide to be integrated in the rest of the code:
     - `input(form)`: To return the html code to be added to the form
     - `apply!(klass)`: Called when enabled. It adds attributes to `klass`, which is a Model
     - `probability(object)`: Returns a number between `0` and `1` on how likely is the client a robot according to the check

### Changes introduced in the design
- Business Logic:
  - `Protector`
    - Now attached to a controller, not a model.
    - Is much simpler, it doesn't include the `FormProtector` class anymore, which has been moved to its own file.
    - `form` method has been moved to `Integration::Controller`
    - It behaves the same way, implements a `spam?` method which run the check challenges.
  - `Configuration`
    - Now attached to a controller, not a model.
    - Also behaves the same way as before, but it's simpler as there's no need to call `apply!` on checks anymore.
    - The unused and thread-unsafe store has been removed
- Integration:
  - Models
    -  `Integration::Model` has been removed and models are completely unrelated to spam protection now. It all happens at in the controller.
  - Controllers
    - `Integration::Controller` is now the center of all the spam protection logic. It includes:
      - A `spam_protection` instance attribute of type `Protector`
      - A `spam_protection_conf` instance attribute of type `Configuration`
      - A helper `spam_protection_form` to return a `FormProtector` which includes the checks or the captcha
      - A single store which is for now implemented by a cookie
      - It's configured by calling `has_spam_protection` on the controller
      - A method `instantiate_checks` is called before any action in order to initialize the spam  checks
      - `spam_check` and `spam_check_save` methods to be called on `PUT` and `POST` request to detect bots and update `marked_as_possible_spam_until`
-  Forms
   - Work the same way
- Checks
  - Instead of reading model attributes, they now read request parameters
  - `apply!` has been removed, there's no need to add model attributes anymore
  - There's a problem with the average system, the more checks exist, the more of them must fail in order to add up an average `> 0.4`. For example, currently the signup screen will detect a bot when javascript is disabled and the form is filled in less than ten seconds, but if we add a couple more checks, then exactly the same behavior will pass, just because there are more tests to count when computing the average. To prevent this, now a check has several possible results:
    - Pass: when it's sure the client is not spam
      - Adds `0` to the average
    - Fail: when it thinks the client is a bot
      - Adds `1` to the average
    - Add any other number to the average
    - Raise error: When it's sure it's a bot, an exception prevents running anymore checks, the average is ignored, and the client is marked as bot.

### How to review
I think it'll be easier to review each commit separately:
- Main refactor: https://github.com/3scale/porta/pull/3620/commits/763b1ed3edb304a9370696e26f053ec87e6ab88e:
  - Moves the spam check logic from the model to the controller
  - Also other changes I mentioned above
  - Update some other files to adapt to the new design
- Allow checks to mark as spam immediately https://github.com/3scale/porta/pull/3620/commits/5ffa0b96b4692d4694feabb69c4f0033dd5ae561:
  - To solve the problem with averages I mentioned above: same behavior being considered spam or not depending on the number of checks enabled
  - Honeypot always raises the exception when failing
  - Other checks raise the exception when anything unexpected happens
- New spam check: Attempts counter https://github.com/3scale/porta/pull/3620/commits/3d23ecaab6971e526d35b33fa41c71377a84c302:
  - Fails when the form is sent three times in less than a minute
  - As Honeypot, it always raises the exception on failing
- Update tests https://github.com/3scale/porta/pull/3620/commits/0a4f98e9b0f47c9cdcf897cf7e59fc900c661eff:
  - Before this PR, there was one big test suite to test all spam check logic ([code](https://github.com/3scale/porta/blob/3scale-2.13.2-GA/test/unit/three_scale/spam_protection_test.rb)). Now tests are split into many files, each one matching the path for the file it's testing.
  - Some integration tests created for `Integration::Controller` and `Integration::FormBuilder`
- Remove all tests for forums https://github.com/3scale/porta/pull/3620/commits/956db657bc3100d3269df91d1775ebf2e4b3608b:
  - Forums had the spam protection enabled on their models. This changes broke many tests for the forum feature. Since forums are going to be removed soon, there's no point on spending time fixing those tests. I simply removed them.
- Add spam protection to the signup screens https://github.com/3scale/porta/pull/3620/commits/758fbfeba9f723021e849643b4867d7de18c698b:
  - Enable `:javascript`, `:honeypot`, `:timestamp` and `:count` challenges to the signup controllers
    - To sign up buyers from the developer portal
    - To sign up providers from 3scale.net
- Add spam protection to the reset password screen https://github.com/3scale/porta/pull/3620/commits/71a2e4cc4f2643284ac9f64f5894af9dde1e2140:
  - Enable `:javascript`, `:honeypot` and `:count` challenges to the reset password form in the developer portal.

### Pending

Check this Jira [comment](https://issues.redhat.com/browse/THREESCALE-9492?focusedId=23480485&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-23480485)

mayorova commented on app/lib/three_scale/spam_protection/checks/base.rb:
```diff
@@ -1,31 +1,51 @@
+    HIDE_STYLE = "position: absolute; width: 1px; height: 0; left: 0; overflow: hidden;"
+    SPAM_CHECKS_SECRET_KEY = -> { Rails.application.key_generator.generate_key('spam-protection-checks', 32) }
+
+    class SpamCheckError < StandardError; end
```
I'm thinking maybe the error name should be more explicit?
Here one can think that some error happened when checking for spam, but we maybe we could call it `SpamDetectedError` or something like that to make it clear that it is thrown when we detect there was an spam attempt.

jlledom replied:
It makes sense. Fixed

mayorova commented on app/lib/three_scale/spam_protection/integration/controller.rb:
```diff
@@ -1,27 +1,74 @@
-            session_store.new(request.session).mark_possible_spam
+          if spam?
+            @store.mark_possible_spam
             Rails.logger.debug "[SpamProtection][Integration] Captcha filled and object is spam - verifying captcha"
```
I know it has been already like that, but I think this log is not always true... 
It is printed even when there was no captcha shown to the user, but the spam checks detected a risk. But well, it's OK.

jlledom replied:
Yes, it's a bit confusing. In `:auto` mode two things can happen: receiving the check inputs, or not receiving them because a captcha was shown in their place. In both cases `spam?` returns `true`. So the logic is correct, but the message is confusing. I reworded it to `Spam detected or captcha filled - verifying captcha`. I hope that's better.

mayorova commented:
Overall, it looks great! Makes much more sense to have it enabled on the controller, rather than on the model.
From what I can tell, it works well also :+1: 

I'll check a bit more to be sure, but I think it's good. Well done @jlledom :clap:

mayorova commented on app/lib/three_scale/spam_protection/integration/controller.rb:
```diff
@@ -1,27 +1,74 @@
       end
 
       def verify_captcha(object)
         verify_recaptcha(model: object, attribute: :recaptcha)
```
Not relevant, as this part has not changed, but from what I saw, it doesn't really make much sense to pass the model and the attribute to `verify_recaptcha`, as even though it adds an error to the model passed to it, we don't actually show this error anyway :upside_down_face: 

in fact, it could make more sense to call this without arguments, so a flash message would be shown.
The downside of it is that on the password reminder form there is another flash message, so there would be redundant flashes (one with proper "error" red style, and another "normal").

But anyway, this does not need to be touched in this PR, it's just something I noticed.

jlledom replied:
You're right, it's better to remove that `object` param if it's being ignored anyway. So we simplify the logic which is complicated enough already. I tried locally and it causes a little mess on the UI which needs time to work on. So better leave it for another PR

mayorova replied:
Yeah, I think there are several things that could be improved, but indeed, it's better we don't change the actual behavior here.

akostadinov commented on app/lib/three_scale/spam_protection/checks/base.rb:
```diff
@@ -1,31 +1,49 @@
+      private
+
+      def encryptor
+        @encryptor ||= ActiveSupport::MessageEncryptor.new(SPAM_CHECKS_SECRET_KEY.call)
```
We will have different encryptors for different workers, correct? How does this affect the logic?

akostadinov reviewed (commented):
Just a preliminary review questions as I need more time to orient myself. 

> The unused and thread-unsafe store has been removed
Could you explain what do you consider thread unsafe in the old store?


> It also manages a store where checks can store data. Note this store is not the SessionStore attached to the cookie. This another store, merely a class attribute, which is therefore one single hash accessible from all the app, and fortunately unused.

If that store was never used, then why do we need to keep it?

mayorova commented on features/step_definitions/signup_steps.rb:
```diff
@@ -87,15 +87,22 @@ def readonly_field(locator)
+  page.evaluate_script <<-JS
+    document.getElementById("#{id}").checked = true
+  JS
+end
```
No empty line :eyes:

mayorova commented on features/support/parameter_types.rb:
```diff
@@ -421,3 +421,9 @@ def find_metric(metrics, name)
+  name: 'how',
+  regexp: /wrong|suspiciously|correctly|again|/,
+  transformer: ->(value) { value }
+)
```
empty line :eyes:

mayorova commented on app/lib/three_scale/spam_protection/form_protector.rb:
```diff
@@ -0,0 +1,65 @@
+
+    # Reads the store to decide whether to show the captcha or not.
+    #
+    # The value is saved in the store from Protector#spam_check on a previous PU or POST request.
```
```suggestion
    # The value is saved in the store from Protector#spam_check on a previous PUT or POST request.
```

jlledom replied:
Fixed

mayorova commented:
I'm wondering whether we really need to print all `[SpamProtection]` logs at `:info` level, maybe `:debug` would be enough?

jlledom commented:
> > The unused and thread-unsafe store has been removed

> Could you explain what do you consider thread unsafe in the old store?

In the old design there were two stores: the cookie, and this thread-unsafe store which was never used. Each protected model had a `Configuration` instance attached the model **class**, so the same `Configuration` instance was being initialized once during application boot and used during all runtime. That unique and application-wide instance had a `@store` hash attribute, shared between all threads, all controllers, etc. with no access protection or anything. In addition, it was completely unused. That's the store I removed, but we keep the cookie store.

> > It also manages a store where checks can store data. Note this store is not the SessionStore attached to the cookie. This another store, merely a class attribute, which is therefore one single hash accessible from all the app, and fortunately unused.

> If that store was never used, then why do we need to keep it?
We don't keep it, I removed it.

jlledom commented:
> I'm wondering whether we really need to print all `[SpamProtection]` logs at `:info` level, maybe `:debug` would be enough?

Right, moved to `:debug`

jlledom commented on features/support/parameter_types.rb:
```diff
@@ -421,3 +421,9 @@ def find_metric(metrics, name)
 )
+
+ParameterType(
+  name: 'how',
```
I made some changes: https://github.com/3scale/porta/pull/3620/commits/c794a966ac57078a39fac319e146be3386b57de4

akostadinov commented on app/lib/three_scale/spam_protection/session_store.rb:
```diff
@@ -1,17 +1,18 @@
 
     def marked_as_possible_spam?
-      session[:marked_as_possible_spam_until].to_i > Time.now.utc.to_i
+      @session[:marked_as_possible_spam_until].to_i > Time.now.utc.to_i
```
Shouldn't this value be encrypted?

jlledom commented:
Good news!

I'm closing this because we decided to simplify all the spam protection logic and probably move to reCaptcha v3.
